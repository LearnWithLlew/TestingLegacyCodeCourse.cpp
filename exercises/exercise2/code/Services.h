#pragma once

#include <iostream>
#include <sstream>

#include "third_party_code.h"

#ifdef LOGGING_TEST
std::stringstream logger;
#endif

int RestCall(std::string url, std::string action, int key)
{
	#ifdef LOGGING_TEST
    logger << "RestCall: url:" << url << "," << "action:" << action << "," << "key:" << key << std::endl;
    return 200;
	#endif
    int result =_third_party_code_::rest_call(url, action, key);
    return result;
}

void initialiseServices()
{
    int key = 42;
    int result = RestCall("https://postgresdatabase.acme.com", "start", key);
    if ( result == 200 )
    {
        key = key + 100;
        result = RestCall("https://kafka.acme.com", "subscribe", key);
        if ( result == 200 )
        {
            result = RestCall("https://cache.acme.com", "load", key);
            if ( result != 200 )
            {
                std::cout << "Error starting cache\n";
            }
            result = RestCall("https://search.acme.com", "index", key);
            if ( result != 200 )
            {
                std::cout << "Error indexing search\n";
            }
            result = RestCall("https://creditcardprocessor.acme.com", "charge", key);
            if ( result != 200 )
            {
                std::cout << "Error charging credit card\n";
            }
        }
    }
}
