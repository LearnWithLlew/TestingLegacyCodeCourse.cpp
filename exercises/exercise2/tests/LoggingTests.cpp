#include <catch2/catch.hpp>
#include "ApprovalTests.hpp"

#define SECURE 1
#define LOGGING_TEST
#include "Services.h"

#include <iostream>

using namespace ApprovalTests;

TEST_CASE("Test initialiseServices")
{
	logger.clear();
	initialiseServices();
	Approvals::verify(logger.str());
}

TEST_CASE("Test initialiseServices 2")
{
	logger.clear();
	initialiseServices();
	Approvals::verify(logger.str());
}
