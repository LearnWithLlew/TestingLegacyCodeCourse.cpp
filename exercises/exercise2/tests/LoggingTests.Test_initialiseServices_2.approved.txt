RestCall: url:https://postgresdatabase.acme.com,action:start,key:42
RestCall: url:https://kafka.acme.com,action:subscribe,key:142
RestCall: url:https://cache.acme.com,action:load,key:142
RestCall: url:https://search.acme.com,action:index,key:142
RestCall: url:https://creditcardprocessor.acme.com,action:charge,key:142

